### ISSUE TITLE
DOC: `memmap` function should reference a serialization protocol.

### ISSUE THREAD

**Issue by ctb** [numpy/numpy#22213]
### Issue with current documentation:

Background:

we use `numpy.memmap` over in sourmash with an array of custom objects (`sourmash.SourmashSignature`) for multiprocessing, and we ran into some problems caused by a bug in our own serialization protocol (https://github.com/sourmash-bio/sourmash/pull/2265). Not numpy's fault at all!

The Problem:

However, in tracking this bug down, I struggled to figure out what serialization protocol was used by memmap, and couldn't find any details at all - serialization isn't even mentioned!

### Idea or request for content:

I think that the memmap documentation should mention serialization, along with whatever pickle protocol is actually being used for complex objects. It's not 100% clear to me where serialization is actually happening in this case - is it perhaps in ndarray? - or else I'd try writing something myself.

Specifically,
* mentioning serialization, however briefly, would be great! It's implicit in memmap if you have any idea of memmap must work, but it took me a while to get there.
* briefly describing how complex objects are pickled/unpickled would also be good, since it doesn't seem to use pickle.load?

**Comment by rkern** [comment]
You mean that you use `np.memmap(..., dtype=object)`? There is no serialization of the objects' values in the memory-mapped file in that case. The only thing in the file is the `PyObject*` pointer values, as that's what's actually in the `dtype=object` array memory.

**Comment by ctb** [comment]
thanks @rkern - the code is here,

https://github.com/sourmash-bio/sourmash/blob/latest/src/sourmash/np_utils.py

and the array that is passed in is created [here](https://github.com/sourmash-bio/sourmash/blob/latest/src/sourmash/compare.py#L239),

so technically it's an ndarray with dtype=object, with objects from class `SourmashSignature`.

And when it's passed to another process in multiprocessing, `__getstate__` is being called on `SourmashSignature`, and then (I think) `__new__` is being used to create the copy of the object. That is why I was pretty sure serialization was being used somewhere.

Apologies if I got this all wrong, I will dig a bit more to figure out what's going on!

**Comment by ctb** [comment]
huh. looks like I was completely wrong and it's multiprocessing that's doing it. My apologies!

**Comment by rkern** [comment]
Yes, `multiprocessing` is pickling the `ndarray` that you gave it to reconstruct on the other side. That it happens to be a `memmap` (an `ndarray` subclass) is mostly a red herring. `memmap` doesn't override anything about pickling the `ndarray`. The reconstructed object won't actually use memory-mapping on the other side (check the `.filename` attribute; it should be `None`).

`ndarray.__reduce_ex__()` will produce a tuple of more primitive objects (including a list with the `SourmashSignature` objects for the values) that `pickle` will then recursively pickle.

**Comment by ctb** [comment]
thanks, much appreciated ;). I've updated the issue and PR over on the sourmash repo side to make this clearer!

**Comment by rkern** [comment]
I don't think `to_memmap()` is helping you here for `dtype=object` arrays. All of the actual `SourmashSignature` objects are still there in your memory (and in the child processes); they won't actually get paged out, just some of the `PyObject*` pointers.

**Comment by ctb** [comment]
yep, we've got another PR about that - I'll be looking into it now that I understand a bit more of what's going on (I wasn't the original author, which was part of the confusion for me!)
